src/servers/src/prom_row_builder.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::string::ToString;
+
+use api::prom_store::remote::Sample;
+use api::v1::value::ValueData;
+use api::v1::{
+    ColumnDataType, ColumnSchema, Row, RowInsertRequest, RowInsertRequests, Rows, SemanticType,
+    Value,
+};
+use common_query::prelude::{GREPTIME_TIMESTAMP, GREPTIME_VALUE};
+
+use crate::proto::PromLabel;
+use crate::repeated_field::Clear;
+
+#[derive(Default)]
+pub(crate) struct TablesBuilder {
+    tables: HashMap<String, TableBuilder>,
+}
+
+impl Clear for TablesBuilder {
+    fn clear(&mut self) {
+        self.tables.clear();
+    }
+}
+
+impl TablesBuilder {
+    pub(crate) fn get_or_create_table_builder(&mut self, table_name: String) -> &mut TableBuilder {
+        self.tables.entry(table_name).or_default()
+    }
+
+    pub(crate) fn as_insert_requests(&mut self) -> (RowInsertRequests, usize) {
+        let mut total_rows = 0;
+        let inserts = self
+            .tables
+            .drain()
+            .map(|(name, mut table)| {
+                total_rows += table.num_rows();
+                table.as_row_insert_request(name)
+            })
+            .collect();
+        (RowInsertRequests { inserts }, total_rows)
+    }
+}
+
+pub(crate) struct TableBuilder {
+    schema: Vec<ColumnSchema>,
+    rows: Vec<Row>,
+    col_indexes: HashMap<String, usize>,
+}
+
+impl Default for TableBuilder {
+    fn default() -> Self {
+        let col_indexes = [
+            (GREPTIME_TIMESTAMP.to_string(), 0),
+            (GREPTIME_VALUE.to_string(), 1),
+        ]
+        .into_iter()
+        .collect();
+        Self {
+            schema: vec![
+                ColumnSchema {
+                    column_name: GREPTIME_TIMESTAMP.to_string(),
+                    datatype: ColumnDataType::TimestampMillisecond as i32,
+                    semantic_type: SemanticType::Timestamp as i32,
+                    datatype_extension: None,
+                },
+                ColumnSchema {
+                    column_name: GREPTIME_VALUE.to_string(),
+                    datatype: ColumnDataType::Float64 as i32,
+                    semantic_type: SemanticType::Field as i32,
+                    datatype_extension: None,
+                },
+            ],
+            rows: vec![],
```

Comment:
I think we know the rows number here, so let's create a `with_capacity` function for  `TableBuilder` and use it in `PromTimeSeries#add_to_table_data`.

Reply:
Added an `TableBuilder::with_capacity` method and  use this method to create the default table builder https://github.com/GreptimeTeam/greptimedb/pull/3425/commits/571926fd0e2cda3f14633107fa722bf6e73c55c8#diff-7c0d742ac3a9cfc88c110b929c9f0a43f63cf981264415ff10c25e0710749ed7R87